### PR TITLE
cilium: bpf-based hostport implementation

### PR DESCRIPTION
**Pull request by borkmann**
See commit msg.

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10592)

### CODE REVIEW

File: cilium/cmd/service_update.go [hunk 1]
```diff
@@ -97,12 +106,14 @@ func updateService(cmd *cobra.Command, args []string) {
 		spec.Flags = &models.ServiceSpecFlags{}
 	}
 
-	if k8sExternalIPs && k8sNodePort {
-		Fatalf("Cannot set both --k8s-external and --k8s-node-port for a service")
+	if boolToInt(k8sExternalIPs)+boolToInt(k8sNodePort)+boolToInt(k8sHostPort) > 1 {
+		Fatalf("Cannot set more than one of --k8s-external, --k8s-node-port, --k8s-host-port for a service")
```
Comment (tklauser) [inline cilium/cmd/service_update.go]: Nit, maybe rephrase this to:

```suggestion
		Fatalf("Can only set one of --k8s-external, --k8s-node-port, --k8s-host-port for a service")
```

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
Prior `Cilium-Ginkgo-Tests` run was green.

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
@borkmann 
```
# github.com/cilium/cilium/test/k8sT
test/k8sT/Services.go:636:26: cannot use httpHostPort (type int) as type int32 in argument to getHTTPLink
test/k8sT/Services.go:637:26: cannot use tftpHostPort (type int) as type int32 in argument to getTFTPLink
test/k8sT/Services.go:646:25: cannot use httpHostPort (type int) as type int32 in argument to getHTTPLink
test/k8sT/Services.go:647:25: cannot use tftpHostPort (type int) as type int32 in argument to getTFTPLink
make[1]: *** [govet] Error 2
make[1]: Leaving directory `/home/travis/gopath/src/github.com/cilium/cilium'
make: *** [unit-tests] Error 2
```

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 2]
```diff
@@ -321,6 +322,92 @@ then the same range must be passed to Cilium via the ``global.nodePort.range``
...
+        --set global.kubeProxyReplacement=strict \\
+        --set global.k8sServiceHost=API_SERVER_IP \\
+        --set global.k8sServicePort=API_SERVER_PORT
+
+The following modified example yaml from the setup validation with an additional
+``hostPort: 8080`` parameter can be used to check the mapping:
+
+.. parsed-literal::
+
+    apiVersion: apps/v1
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst:373]: Nit: `kubectl run nginx --generator=run-pod/v1 --image=nginx --replicas=1 --port=80 --hostport=8080`?

File: Documentation/gettingstarted/kubeproxy-free.rst [hunk 3]
```diff
@@ -321,6 +322,92 @@ then the same range must be passed to Cilium via the ``global.nodePort.range``
...
+    kind: Deployment
+    metadata:
+      name: my-nginx
+    spec:
+      selector:
+        matchLabels:
+          run: my-nginx
+      replicas: 2
```
Comment (brb) [inline Documentation/gettingstarted/kubeproxy-free.rst]: What happens if you expose two pods on the same host on the same hostPort?

File: install/kubernetes/cilium/values.yaml [hunk 4]
```diff
@@ -263,6 +263,11 @@ global:
     # mode is the mode of NodePort feature
     mode: "hybrid"
 
+  # hostPort is the configuration for container hostPort mapping
+  hostPort:
+    # enabled enables the hostPort functionality
+    enabled: false
```
Comment (brb) [inline install/kubernetes/cilium/values.yaml:269]: Do we want to disable the hostPort feature for new users? Asking, as for the existing users after the upgrade to v1.8 the feature will be enabled due to the agent defaulting the flag to `true`.

**Review by brb — APPROVED** [review]
Nice! Couple nits and questions.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 3 above)
Comment (borkmann) [reply Documentation/gettingstarted/kubeproxy-free.rst]: The second deployment remains in `Pending`: 
```
Events:
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  0/1 nodes are available: 1 node(s) didn't have free ports for the requested pod ports.
```
And once the first deployment is deleted, then it can proceed and also create the new HostPort service along with that.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 2 above)
Comment (borkmann) [reply Documentation/gettingstarted/kubeproxy-free.rst:373]: I was referring to the example yaml from earlier in the same doc, so I kept it as is but with the added `hostPort`.

File: install/kubernetes/cilium/values.yaml (hunk 4 above)
Comment (borkmann) [reply install/kubernetes/cilium/values.yaml:269]: I would keep it enabled otherwise people might miss it and still deploy the `portmap` CNI chaining. I'll do a follow-up for the upgrade guide to document the support and actions needed.

File: Documentation/gettingstarted/kubeproxy-free.rst (hunk 3 above)
Comment (brb) [reply Documentation/gettingstarted/kubeproxy-free.rst]: Maybe then it makes sense to decrease the `replicas` count to 1, as the `Pending` state might confuse users who want to validate whether the feature works.

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-docs-please

**Comment by borkmann** [comment]
(prior docs test was green)

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
(prior travis ci was green, x86 is green here, arm64 flaky in unrelated spot)

**Comment by borkmann** [comment]
test-me-please
